pulsar-client/src/main/java/org/apache/pulsar/client/impl/RoundRobinPartitionMessageRouterImpl.java

```diff
@@ -79,9 +69,15 @@ public int choosePartition(Message<?> msg, TopicMetadata topicMetadata) {
             return signSafeMod(hash.makeHash(msg.getKey()), topicMetadata.numPartitions());
         }
 
-        if (isBatchingEnabled) { // if batching is enabled, choose partition on `maxBatchingDelayMs` boundary.
-            long currentMs = clock.millis();
-            return signSafeMod(currentMs / maxBatchingDelayMs + startPtnIdx, topicMetadata.numPartitions());
+        if (isBatchingEnabled) {
+            int msgSize = msg.getData().length;
+            if (msgSize + MESSAGE_BYTES_UPDATER.get(this) <= maxBatchingBytes) {
```

Comment:
I don't think we need to be very accurate here. we should just use addAndGet directly.

```
int newBatchSize = MESSAGE_BYTES_UPDATER.addAndGet(this, msg.getData().length);
if (newBatchSize <= maxBatchingBytes) {
   return signSafeMod(PARTITION_INDEX_UPDATER.get(this), topicMetadata.numPartitions());
} else {
   MESSAGE_BYTES_UPDATER.set(this, 0);
   return signSafeMod(PARTITION_INDEX_UPDATER.getAndIncrement(this), topicMetadata.numPartitions());
}
```

Reply:
Actually I think this is still very tricky, because we might skip partitions when there are concurrent accesses to the router.